File: examples/kinesis/0_basic/1_update_did.rs
// Copyright 2020-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use examples_kinesis::create_kinesis_did_document;
use examples_kinesis::get_client_and_create_account;
use examples_kinesis::get_memstorage;
use examples_kinesis::TEST_GAS_BUDGET;
use identity_iota::core::json;
use identity_iota::core::FromJson;
use identity_iota::core::Timestamp;
use identity_iota::did::DIDUrl;
use identity_iota::did::DID;
use identity_iota::document::Service;
use identity_iota::iota::IotaDID;
use identity_iota::iota::IotaDocument;
use identity_iota::storage::JwkDocumentExt;
use identity_iota::storage::JwkMemStore;
use identity_iota::verification::jws::JwsAlgorithm;
use identity_iota::verification::MethodRelationship;
use identity_iota::verification::MethodScope;
use identity_storage::StorageSigner;

/// Demonstrates how to update a DID document in an existing Alias Output.
#[tokio::main]
async fn main() -> anyhow::Result<()> {
  // create new client to interact with chain and get funded account with keys
  let storage = get_memstorage()?;
  let (identity_client, key_id, public_key_jwk) = get_client_and_create_account(&storage).await?;
  // create new signer that will be used to sign tx with
  let signer = StorageSigner::new(&storage, key_id, public_key_jwk);
  // create new DID document and publish it
  let (document, vm_fragment_1) = create_kinesis_did_document(&identity_client, &storage, &signer).await?;
  let did: IotaDID = document.id().clone();

  // Resolve the latest state of the document.
  let mut document: IotaDocument = identity_client.resolve_did(&did).await?;

  // Insert a new Ed25519 verification method in the DID document.
  let vm_fragment_2: String = document
    .generate_method(
      &storage,
      JwkMemStore::ED25519_KEY_TYPE,
      JwsAlgorithm::EdDSA,
      None,
      MethodScope::VerificationMethod,
    )
    .await?;

  // Attach a new method relationship to the inserted method.
  document.attach_method_relationship(
    &document.id().to_url().join(format!("#{vm_fragment_2}"))?,
    MethodRelationship::Authentication,
  )?;

  // Add a new Service.
  let service: Service = Service::from_json_value(json!({
    "id": document.id().to_url().join("#linked-domain")?,
    "type": "LinkedDomains",
    "serviceEndpoint": "https://iota.org/"
  }))?;
  assert!(document.insert_service(service).is_ok());
  document.metadata.updated = Some(Timestamp::now_utc());

  // Remove a verification method.
  let original_method: DIDUrl = document.resolve_method(&vm_fragment_1, None).unwrap().id().clone();
  document.purge_method(&storage, &original_method).await.unwrap();

  let updated = identity_client
    .publish_did_document_update(document.clone(), TEST_GAS_BUDGET, &signer)
    .await?;

  println!("Updated DID document: {updated:#}");

  Ok(())
}

File: bindings/wasm/identity_wasm/examples/src/0_basic/1_update_did.ts
// Copyright 2020-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    JwkMemStore,
    JwsAlgorithm,
    MethodRelationship,
    MethodScope,
    Service,
    VerificationMethod,
} from "@iota/identity-wasm/node";
import { IotaClient } from "@iota/iota-sdk/client";
import { createDocumentForNetwork, getFundedClient, getMemstorage, NETWORK_URL, TEST_GAS_BUDGET } from "../util";

/** Demonstrates how to update a DID document in an existing identity. */
export async function updateIdentity() {
    // create new clients and create new account
    const iotaClient = new IotaClient({ url: NETWORK_URL });
    const network = await iotaClient.getChainIdentifier();
    const storage = getMemstorage();
    const identityClient = await getFundedClient(storage);
    const [unpublished, vmFragment1] = await createDocumentForNetwork(storage, network);

    // create new identity for this account and publish document for it
    const { output: identity } = await identityClient
        .createIdentity(unpublished)
        .finish()
        .buildAndExecute(identityClient);
    const did = identity.didDocument().id();

    // Resolve the latest state of the document.
    // Technically this is equivalent to the document above.
    const resolved = await identityClient.resolveDid(did);

    // Insert a new Ed25519 verification method in the DID document.
    const vmFragment2 = await resolved.generateMethod(
        storage,
        JwkMemStore.ed25519KeyType(),
        JwsAlgorithm.EdDSA,
        null,
        MethodScope.VerificationMethod(),
    );

    // Attach a new method relationship to the inserted method.
    resolved.attachMethodRelationship(did.join(`#${vmFragment2}`), MethodRelationship.Authentication);

    // Add a new Service.
    const service: Service = new Service({
        id: did.join("#linked-domain"),
        type: "LinkedDomains",
        serviceEndpoint: "https://iota.org/",
    });
    resolved.insertService(service);

    // Remove a verification method.
    let originalMethod = resolved.resolveMethod(vmFragment1) as VerificationMethod;
    await resolved.purgeMethod(storage, originalMethod?.id());

    let controllerToken = await identity.getControllerToken(identityClient);

    let maybePendingProposal = await identity
        .updateDidDocument(resolved.clone(), controllerToken!)
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(identityClient)
        .then(result => result.output);

    console.assert(maybePendingProposal == null, "the proposal should have been executed right away!");

    // and resolve again to make sure we're looking at the onchain information
    const resolvedAgain = await identityClient.resolveDid(did);
    console.log(`Updated DID document result: ${JSON.stringify(resolvedAgain, null, 2)}`);
}
